Add directive @bool to Blade

david-valdivia opened the pull request:
Add @bool directive functionality to Blade, allowing boolean values to be printed directly into strings or used in object construction.

Examples: 

**JS**

```
<script>
    let config = {
        isActive: @bool($isActive),
        hasAccess: @bool($hasAccess)
    };
</script>
```

**Alpine**

```
<div x-data="{ isActive: @bool($isActive) }">
    <button :class="{ 'active': isActive }">Toggle</button>
</div>
```

Bootstrap:
```
<div class="dropdown">
  <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="@bool($hasPopup)" aria-expanded="@bool($isExpanded)">
    Dropdown button
  </button>
  <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
    <a class="dropdown-item" href="#">Action</a>
    <a class="dropdown-item" href="#">Another action</a>
    <a class="dropdown-item" href="#">Something else here</a>
  </div>
</div>
```

david-valdivia committed:
Add directive @bool to Blade

Add @bool directive functionality to Blade, allowing boolean values to be printed directly into strings or used in object construction.

bert-w commented on src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php:
```diff
@@ -316,6 +316,17 @@ protected function compileSelected($condition)
+     */
+    protected function compileBool($condition)
+    {
+        return "<?php if{$condition}: echo 'true'; else: 'false'; endif; ?>";
```
Your code doesn't print anything for the "false" path.

return  "<?php echo {$condition} ? 'true' : 'false'; ?>";

bert-w commented on tests/View/Blade/BladeBoolTest.php:
```diff
@@ -0,0 +1,30 @@
+
+class BladeBoolTest extends AbstractBladeTestCase
+{
+    public function testBool()
```
This is probably one of the few cases where an `eval()` test would have found the error above, like so:

```php
public function testCompileBool(): void
    {
        $someViewVarTruthy = 123;
        $compiled = $this->compiler->compileString('@bool($someViewVarTruthy)');

        ob_start();
        eval(substr($compiled, 6, -3));
        $this->assertEquals('true', ob_get_clean());

        $someViewVarFalsey = '0';
        $compiled = $this->compiler->compileString('@bool($someViewVarFalsey)');

        ob_start();
        eval(substr($compiled, 6, -3));
        $this->assertEquals('false', ob_get_clean());

    }
```

dvaldiviacom replied:
You're right, it didn't consider a Falsey values, i added a Falsey support and a couple of test for NULL and Class instances.

david-valdivia committed:
Fix @bool suggestion

Fix errors that occours when a Falsey value like '0' is passed, also added a NULL and instance testing.

david-valdivia committed:
continuous-integration

david-valdivia committed:
Update BladeBoolTest.php

david-valdivia committed:
Update BladeBoolTest.php

taylorotwell committed:
Update CompilesConditionals.php

taylorotwell committed:
Update CompilesConditionals.php

MrPunyapal commented:
btw it works with `@js(true)` out of the box.

ahinkle commented:
What's the difference between `@js` and `@bool`? Would you care to contribute to the docs?

david-valdivia commented:
@MrPunyapal @ahinkle 
> @david-valdivia What's the difference between `@js($condition)` and `@bool($condition)`? Would you care to contribute to the docs?

no problem, \@bool is thinked only in print a boolean values, means that it could recognize between Falsey and Truthy values o diference with \@js that convert into json.

Cases:
<img width="518" alt="Captura de pantalla 2024-10-29 a la(s) 10 33 54 a m" src="https://github.com/user-attachments/assets/c59dbbca-336e-451f-820f-5bebfb59caba">

Results:
<img width="309" alt="Captura de pantalla 2024-10-29 a la(s) 10 34 55 a m" src="https://github.com/user-attachments/assets/cca60aae-7355-40d1-af26-4b7e31724ade">

\@bool() always returns a boolean value, @js() converts into a json.
